File: cpp/src/ladon.cpp
#include <fstream>
#include <iostream>
#include <vector>

#include "./scanner.cpp"

void runFile(const char* path);
void runPrompt();
void run(const char* source);
void error(int line, const char* message);
void report(int line, const char* where, const char* message);

bool hadError = false;

int main(int argc, char const* argv[]) {
  if (argc > 2) {
    std::cout << "Usage: ladon [script]" << std::endl;
  } else if (argc == 2) {
    runFile(argv[1]);
  } else {
    runPrompt();
  }

  return 0;
}

void runFile(const char* path) {
  std::ifstream file(path, std::ios::binary | std::ios::ate);
  if (!file) {
    std::cout << "Could not open file" << std::endl;
    exit(74);
  }

  std::streamsize size = file.tellg();
  // move to the beginning of the file
  file.seekg(0, std::ios::beg);

  char* buffer = new char[size];

  // read the content of the file
  file.read(buffer, size);

  std::cout << buffer << std::endl;

  run(buffer);

  if (hadError) exit(65);
}

void run(const char* source) {
  Scanner scanner(source);
  std::vector<Token> tokens = scanner.scanTokens();

  /*for (Token token : tokens) {
    std::cout << token << std::endl;
  }*/
}

void error(int line, const char* message) { report(line, "", message); }

void report(int line, const char* where, const char* message) {
  std::cerr << "[line " << line << "] Error" << where << ": " << message
            << std::endl;
  hadError = true;
}

void runPrompt() {
  std::string line;
  while (true) {
    std::cout << "> ";
    std::getline(std::cin, line);
    if (line.empty()) break;
    run(line.c_str());
    hadError = false;
  }

  hadError = false;
}

File: java_source/Ladon/src/main/java/org/marco/Ladon.java
package org.marco;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Ladon {
    static boolean hadError = false;


    public static void main(String[] args) throws IOException {
        //print current working dir
        System.out.println(System.getProperty("user.dir"));

        if (args.length > 1){
            System.out.println("Usage: ladon [script]");
            System.exit(64);
        } else if (args.length == 1) {
            runFile(args[0]);
        } else {
            System.out.println("Welcome to Ladon!");
            if (hadError) System.exit(65);
        }
    }

    private static void runFile(String path) throws IOException{
        //get the input from the file
        byte[] bytes = Files.readAllBytes(Paths.get(path));
        //send the input to the interpreter1
        run(new String(bytes, Charset.defaultCharset()));
    }

    private static void run(String source){
        Scanner scanner = new Scanner(source);
        List<Token> tokens = scanner.scanTokens();

        //print tokens
        for (Token token : tokens){
            System.out.println(token);
        }
    }

    static void error(int line, String message){
        report(line, "", message);
    }

    private static void report(int line, String where, String message){
        System.err.println("[line " + line + "] Error" + where + ": " + message);
    }
}